### PR TITLE
MIR-borrowck: Adding notes to E0506

### PR DESCRIPTION
**Pull request by zilbuz**
This PR adds notes to the MIR borrowck error E0506.

Part of #44596

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Could you please add a test for this change? Take a look at how they're done in `src/test/compile-fail`.

**Comment by zilbuz** [comment]
There is a test for E0506 (https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/E0506.rs), but it doesn't test this message because the compiler uses the AST borrowck instead of the MIR borrowck. I manually checked that the messages were the same though. I assumed that the tests for MIR borrowck would be added on a later PR (@pnkfelix ?). Or maybe you're asking for a different kind of test ?

**Comment by KiChjang** [comment]
We could actually use a UI test for this, but it's not feasible right now because we can't properly display lvalues properly in the MIR borrowck.

**Comment by pnkfelix** [comment]
In principle we *could* add a test, by telling the test infrastructure to pass the necessary `-Z` flags.

In particular, you can do `// compile-flags: -Z emit-end-regions -Z borrowck-mir` at the top of the file. (You can find other tests that pass other such flags in `src/test/compile-fail/`)

(I *assume* that the UI tests accept the same `compile-flags` comment.)

Whether its a *good* idea to put in tests for MIR-borrowck at this stage... it probably is worth it, just to make sure we don't regress. But maybe they should go into their own sub-directory or something, just to keep them isolated from "real tests" of current functionality.

**Comment by pnkfelix** [comment]
(After discussion in rust-impl-period/WG-compiler-nll, we decided that we will try the approach of extending the existing tests under `compile-fail/` via the `// revisions: ...` system so that we can keep a single test file and list different sets of expected error messages based on whether or not we are passing `-Z borrowck-mir`.)

**Comment by zilbuz** [comment]
Alright I edited the tests on compile-fail that raise E0506. I omitted the tests that triggers ICE as other issues are supposed to fix them.

**Comment by arielb1** [comment]
Nice. I suppose we'll also want to add "borrow ends here" when we start supporting these, which should be soon.

@bors r+

File: src/librustc_mir/borrow_check.rs [hunk 1]
```diff
@@ -975,10 +975,19 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
         err.emit();
     }
 
-    fn report_illegal_mutation_of_borrowed(&mut self, _: Context, (lvalue, span): (&Lvalue, Span)) {
+    fn report_illegal_mutation_of_borrowed(&mut self,
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: I think this should be moved to `borrowck_errors` actually, but that could be done in another commit.

Comment (zilbuz) [reply src/librustc_mir/borrow_check.rs]: You mean like @mikhail-m1 has done in #44882 ?

File: src/test/compile-fail/coerce-overloaded-autoderef.rs [hunk 2]
```diff
@@ -17,24 +20,32 @@ fn borrow2<T>(_: &mut T, _: &T) {}
...
+    //[ast]~^ ERROR cannot borrow `*x` as mutable more than once at a time
+    //[mir]~^^ ERROR cannot borrow `*x` as mutable more than once at a time (Ast)
+    //[mir]~| ERROR cannot borrow `(*x)` as mutable more than once at a time (Mir)
 }
 
 fn double_borrow2<T>(x: &mut Box<T>) {
     borrow2(x, x);
-    //~^ ERROR cannot borrow `*x` as immutable because it is also borrowed as mutable
+    //[ast]~^ ERROR cannot borrow `*x` as immutable because it is also borrowed as mutable
+    //[mir]~^^ ERROR cannot borrow `*x` as immutable because it is also borrowed as mutable (Ast)
```
Comment (pnkfelix) [inline src/test/compile-fail/coerce-overloaded-autoderef.rs]: the `~^^` here could just be `~|`, right?

File: src/test/compile-fail/coerce-overloaded-autoderef.rs [hunk 3]
```diff
@@ -17,24 +20,32 @@ fn borrow2<T>(_: &mut T, _: &T) {}
...
+    //[ast]~^ ERROR cannot assign to `**x` because it is borrowed
+    //[mir]~^^ ERROR cannot assign to `**x` because it is borrowed (Ast)
+    //[mir]~| ERROR cannot assign to `(*(*x))` because it is borrowed (Mir)
 }
 
 fn double_mut_borrow2<T>(x: &mut Box<T>) {
     borrow_mut2(x, x);
-    //~^ ERROR cannot borrow `*x` as mutable more than once at a time
+    //[ast]~^ ERROR cannot borrow `*x` as mutable more than once at a time
+    //[mir]~^^ ERROR cannot borrow `*x` as mutable more than once at a time (Ast)
```
Comment (pnkfelix) [inline src/test/compile-fail/coerce-overloaded-autoderef.rs]: the `~^^` here could just be `~|`, right?

File: src/test/compile-fail/coerce-overloaded-autoderef.rs [hunk 4]
```diff
@@ -17,24 +20,32 @@ fn borrow2<T>(_: &mut T, _: &T) {}
...
+    //[mir]~| ERROR cannot borrow `(*x)` as mutable more than once at a time (Mir)
 }
 
 fn double_imm_borrow(x: &mut Box<i32>) {
     let y = borrow(x);
     let z = borrow(x);
     **x += 1;
-    //~^ ERROR cannot assign to `**x` because it is borrowed
+    //[ast]~^ ERROR cannot assign to `**x` because it is borrowed
+    //[mir]~^^ ERROR cannot assign to `**x` because it is borrowed (Ast)
```
Comment (pnkfelix) [inline src/test/compile-fail/coerce-overloaded-autoderef.rs]: the `~^^` here could just be `~|`

File: src/test/compile-fail/coerce-overloaded-autoderef.rs [hunk 5]
```diff
@@ -17,24 +20,32 @@ fn borrow2<T>(_: &mut T, _: &T) {}
 fn double_mut_borrow<T>(x: &mut Box<T>) {
     let y = borrow_mut(x);
     let z = borrow_mut(x);
-    //~^ ERROR cannot borrow `*x` as mutable more than once at a time
+    //[ast]~^ ERROR cannot borrow `*x` as mutable more than once at a time
+    //[mir]~^^ ERROR cannot borrow `*x` as mutable more than once at a time (Ast)
```
Comment (pnkfelix) [inline src/test/compile-fail/coerce-overloaded-autoderef.rs]: the `~^^` here could just be `~|`

File: src/test/compile-fail/borrowck/borrowck-overloaded-index-and-overloaded-deref.rs [hunk 6]
```diff
@@ -39,7 +42,9 @@ fn main() {
     let mut v = MyVec { x: MyPtr { x: Foo { f: 22 } } };
     let i = &v[0].f;
     v = MyVec { x: MyPtr { x: Foo { f: 23 } } };
-    //~^ ERROR cannot assign to `v`
+    //[ast]~^ ERROR cannot assign to `v`
+    //[mir]~^^ ERROR cannot assign to `v` because it is borrowed (Ast)
```
Comment (pnkfelix) [inline src/test/compile-fail/borrowck/borrowck-overloaded-index-and-overloaded-deref.rs]: the `~^^` here could just be `~|`

File: src/test/compile-fail/coerce-overloaded-autoderef.rs (hunk 2 above)
Comment (zilbuz) [reply src/test/compile-fail/coerce-overloaded-autoderef.rs]: No, it seems that when using revisions, the comments of the revision not currently compiled aren't processed. If I use `~|`, I have the following error : 
`compile-fail\coerce-overloaded-autoderef.rs' panicked at 'encountered //~| without preceding //~^ line.', src\libcore\option.rs:839:4`

File: src/test/compile-fail/coerce-overloaded-autoderef.rs (hunk 3 above)
Comment (zilbuz) [reply src/test/compile-fail/coerce-overloaded-autoderef.rs]: Cf. https://github.com/rust-lang/rust/pull/44811#discussion_r141350517

File: src/test/compile-fail/coerce-overloaded-autoderef.rs (hunk 4 above)
Comment (zilbuz) [reply src/test/compile-fail/coerce-overloaded-autoderef.rs]: Cf. https://github.com/rust-lang/rust/pull/44811#discussion_r141350517

File: src/test/compile-fail/coerce-overloaded-autoderef.rs (hunk 5 above)
Comment (zilbuz) [reply src/test/compile-fail/coerce-overloaded-autoderef.rs]: Cf. https://github.com/rust-lang/rust/pull/44811#discussion_r141350517

File: src/test/compile-fail/borrowck/borrowck-overloaded-index-and-overloaded-deref.rs (hunk 6 above)
Comment (zilbuz) [reply src/test/compile-fail/borrowck/borrowck-overloaded-index-and-overloaded-deref.rs]: Cf. https://github.com/rust-lang/rust/pull/44811#discussion_r141350517

File: src/librustc_mir/borrow_check.rs (hunk 1 above)
Comment (arielb1) [reply src/librustc_mir/borrow_check.rs]: Yeah

Comment (zilbuz) [reply src/librustc_mir/borrow_check.rs]: Done here: #44922

**Pull request merged**